src/merkleblock.h

```diff
@@ -154,6 +153,10 @@ class CMerkleBlock
         READWRITE(header);
         READWRITE(txn);
     }
+
+private:
+    // Combined constructor to consolidate code
+    CMerkleBlock(const CBlock& block, CBloomFilter* filter, const std::set<uint256>* txids);
```

Comment:
Instead of pointers, you can make it receive references, just pass either `CBloomFilter()` or `std::set<uint256>()`, and remove the ` != nullptr` conditions above. Note that `CBloomFilter` is optimized when it is empty.

If you do this then this constructor can be public.

Reply:
I don't think this will work in the case of `filter` being unused. I'm getting the following compilation error: 
```
In file included from merkleblock.cpp:6:0:

merkleblock.h: In constructor ‘CMerkleBlock::CMerkleBlock(const CBlock&, const std::set<uint256>&)’:
merkleblock.h:145:93: error: invalid initialization of non-const reference of type ‘CBloomFilter&’ from an rvalue of type ‘CBloomFilter’
     CMerkleBlock(const CBlock& block, const std::set<uint256>& txids) : CMerkleBlock(block, CBloomFilter(), txids) {}
                                                                                             ^~~~~~~~~~~~~~
merkleblock.h:148:5: note:   initializing argument 2 of ‘CMerkleBlock::CMerkleBlock(const CBlock&, CBloomFilter&, const std::set<uint256>&)’
     CMerkleBlock(const CBlock& block, CBloomFilter& filter, const std::set<uint256>& txids);
     ^~~~~~~~~~~~
```

Apparently we can't create an unnamed, non-const `CBloomFilter` reference from an rvalue. I'm not sure how to avoid this when using constructor delegation.

Reply:
You're not allowed to pass a temporary into a method that expects a non-const reference. Non-const references are intended for output arguments, and passing a temporary would mean there is effectively no output argument the caller can observe.

It seems to be that using a pointer which can be optionally `nullptr` is the correct approach here.

Reply:
Should have tested my suggestion. Agree with this approach.